### PR TITLE
added support for TestNG

### PR DESCRIPTION
**Pull request by nitizkumar**
Hi An,

Added test ng support for the plugin. Please review.

### CODE REVIEW

File: core/src/main/java/org/lint/azzert/command/ExemptDisabledTestsCommand.java [hunk 1]
```diff
@@ -12,6 +12,11 @@ public Void execute(final Context context){
         context.getMethodContexts().removeIf(
                 method -> ! Collections.disjoint(context.getSupportedExemptApis(), method.getAnnotations())
         );
+
```
Comment (An-Nadein) [inline core/src/main/java/org/lint/azzert/command/ExemptDisabledTestsCommand.java]: would've made sense to fold method.isIgnored() into the filter above. also, this seems to be a different approach to identifying disabled tests, will take a detailed look later.

File: core/src/main/java/org/lint/azzert/context/Context.java [hunk 2]
```diff
@@ -16,6 +16,7 @@
     private final Set<String> assertApis;
     private final Set<String> testFrameworks;
     private final Set<String> exemptApis;
+    private final Set<String> supportedParameters;
```
Comment (An-Nadein) [inline core/src/main/java/org/lint/azzert/context/Context.java:19]: How/Where are the supported params being used?

File: core/src/main/java/org/lint/azzert/context/TestMethodContext.java [hunk 3]
```diff
@@ -30,14 +31,16 @@ public TestMethodContext(TestMethodContext source) {
...
         this.assertMethodsAtLineNumbers = new LinkedList<>(source.assertMethodsAtLineNumbers);
+        this.ignored = source.ignored;
     }
 
     public void resetMethodDetails() {
         this.methodName = "";
         this.methodSignature = "";
         this.annotations = new HashSet<>();
-        this.assertMethodsAtLineNumbers  = new LinkedList<>();
+        this.assertMethodsAtLineNumbers = new LinkedList<>();
```
Comment (An-Nadein) [inline core/src/main/java/org/lint/azzert/context/TestMethodContext.java]: Could you please honor the existing code format in all places?

File: core/src/main/java/org/lint/azzert/visitor/LintAssertAnnotationVisitor.java [hunk 4]
```diff
@@ -0,0 +1,25 @@
+package org.lint.azzert.visitor;
+
+import org.lint.azzert.context.Context;
+import org.objectweb.asm.AnnotationVisitor;
+
+public class LintAssertAnnotationVisitor extends AnnotationVisitor {
```
Comment (An-Nadein) [inline core/src/main/java/org/lint/azzert/visitor/LintAssertAnnotationVisitor.java:6]: See, in JUnit, methods marked with @Ignore and @Disabled are not considered to be tests unless they are also annotated with @Test. That's why I chose not to rely on the AnnotationVisitor and use the ExemptDisabledTestsCommand to filter out all @Test methods with @Ignore/@Disabled annotation

File: core/src/main/java/org/lint/azzert/visitor/LintAssertMethodVisitor.java [hunk 5]
```diff
@@ -23,9 +23,9 @@ public LintAssertMethodVisitor(Context ctx) {
     public AnnotationVisitor visitAnnotation(String descriptor, boolean visible) {
         if (this.context.getAcceptableMethodAnnotations().contains(descriptor)){
             context.with(descriptor, visible);
+            return new LintAssertAnnotationVisitor(context);
```
Comment (An-Nadein) [inline core/src/main/java/org/lint/azzert/visitor/LintAssertMethodVisitor.java:26]: see the comment above

File: core/src/main/java/org/lint/azzert/visitor/LintAssertMethodVisitor.java [hunk 6]
```diff
@@ -23,9 +23,9 @@ public LintAssertMethodVisitor(Context ctx) {
...
         }
-
-        return super.visitAnnotation(descriptor, visible);
+        return super.visitAnnotation(descriptor,visible);
```
Comment (An-Nadein) [inline core/src/main/java/org/lint/azzert/visitor/LintAssertMethodVisitor.java]: see the styling comment above

File: core/src/test/java/org/lint/azzert/AssertTestNGStyle.java [hunk 7]
```diff
@@ -0,0 +1,34 @@
+package org.lint.azzert;
```
Comment (An-Nadein) [inline core/src/test/java/org/lint/azzert/AssertTestNGStyle.java:1]: Did you mean to test methods with parameters as well? Handling parameters would certainly enrich the output and help differentiate between overloaded methods.

File: core/src/test/java/org/lint/azzert/LintAssertTest.java [hunk 8]
```diff
@@ -46,4 +46,44 @@ void azzert() throws Exception {
         Assertions.assertEquals(9, assertsInMethod.apply(methods, "azzert"));
     }
 
+    @Test
```
Comment (An-Nadein) [inline core/src/test/java/org/lint/azzert/LintAssertTest.java:49]: Nice. Let's add a similar test to the client module - it lacks meaningful tests.

File: build.gradle [hunk 9]
```diff
@@ -37,7 +37,3 @@ allprojects {
         }
     }
 }
-
```
Comment (An-Nadein) [inline build.gradle:40]: thank you, forgot to remove that.

**Review by An-Nadein — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/lint/azzert/visitor/LintAssertAnnotationVisitor.java (hunk 4 above)
Comment (nitizkumar) [reply core/src/main/java/org/lint/azzert/visitor/LintAssertAnnotationVisitor.java:6]: TestNG does not have @ignore but used @Test(enabled=false) , need to use AnnotationVisitor to process the parameters. This approach could also be used to filter out the false negatives of @Test(expected=SomeException.class) in JUnit 4

File: core/src/main/java/org/lint/azzert/context/TestMethodContext.java (hunk 3 above)
Comment (nitizkumar) [reply core/src/main/java/org/lint/azzert/context/TestMethodContext.java]: Sure

File: core/src/main/java/org/lint/azzert/context/Context.java (hunk 2 above)
Comment (nitizkumar) [reply core/src/main/java/org/lint/azzert/context/Context.java:19]: Supported parameters currently holds only "enabled" for testNG, I didnt want to use test ng specific code which is why added a supported parameters, currently supported parameter can only be true or false.

File: core/src/main/java/org/lint/azzert/command/ExemptDisabledTestsCommand.java (hunk 1 above)
Comment (nitizkumar) [reply core/src/main/java/org/lint/azzert/command/ExemptDisabledTestsCommand.java]: is ignored is a flag which is added when enabled is false for TestNG, merging the filters

**Comment by An-Nadein** [comment]
Very many thanks for the PR. I could see how cumbersome it was to introduce a new framework support and refactored the code incorporating your changes in. I will close this PR.

**Pull request closed without merging**